File: libs/solvers/python/tests/test_adapt_mpi.py
# ============================================================================ #
# Copyright (c) 2024 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# ADAPT-VQE MPI work-splitting test.
# Verifies that ADAPT-VQE produces the correct H2 ground-state energy when
# commutator evaluation is distributed across multiple MPI ranks.
#
# Run: mpiexec -np 4 python test_adapt_mpi.py

import sys
import numpy as np
import cudaq
import cudaq_solvers as solvers

cudaq.mpi.initialize()

geometry = [('H', (0., 0., 0.)), ('H', (0., 0., .7474))]
molecule = solvers.create_molecule(geometry, 'sto-3g', 0, 0, casci=True)
operators = solvers.get_operator_pool("spin_complement_gsd",
                                      num_orbitals=molecule.n_orbitals)
numElectrons = molecule.n_electrons


@cudaq.kernel
def initState(q: cudaq.qview):
    for i in range(numElectrons):
        x(q[i])


energy, thetas, ops = solvers.adapt_vqe(initState, molecule.hamiltonian,
                                        operators)

rank = cudaq.mpi.rank()
num_ranks = cudaq.mpi.num_ranks()

if rank == 0:
    print(f"[MPI ADAPT] ranks={num_ranks}, energy={energy:.6f}")
    assert np.isclose(energy, -1.137, atol=1e-3), \
        f"MPI ADAPT energy {energy} does not match expected -1.137"
    assert len(ops) > 0, "Expected at least one operator to be selected"
    print("PASS")

cudaq.mpi.finalize()

File: libs/solvers/unittests/test_adapt_mpi.cpp
/*******************************************************************************
 * Copyright (c) 2024 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// ADAPT-VQE MPI work-splitting test.
// Verifies that ADAPT-VQE produces the correct H2 ground-state energy when
// commutator evaluation is distributed across multiple MPI ranks.
//
// Run: mpiexec -np 4 ./test_adapt_mpi

#include <cassert>
#include <cmath>
#include <iostream>

#include "cudaq.h"
#include "nvqpp/test_kernels.h"
#include "cudaq/solvers/adapt.h"
#include "cudaq/solvers/operators.h"

int main(int argc, char **argv) {
  cudaq::mpi::initialize(argc, argv);

  auto geometryHH = cudaq::solvers::molecular_geometry{{"H", {0., 0., 0.}},
                                                       {"H", {0., 0., .7474}}};
  auto hh = cudaq::solvers::create_molecule(
      geometryHH, "sto-3g", 0, 0,
      {.casci = true, .ccsd = false, .verbose = false});

  auto pool = cudaq::solvers::operator_pool::get("spin_complement_gsd");
  auto poolList =
      pool->generate({{"num-orbitals", hh.hamiltonian.num_qubits() / 2}});

  auto [energy, thetas, ops] =
      cudaq::solvers::adapt_vqe(hartreeFock2Electrons, hh.hamiltonian, poolList,
                                {{"grad_norm_tolerance", 1e-3},
                                 {"max_iter", 15},
                                 {"grad_norm_diff_tolerance", 1e-5},
                                 {"threshold_energy", 5e-6}});

  auto rank = cudaq::mpi::rank();
  auto numRanks = cudaq::mpi::num_ranks();

  if (rank == 0) {
    std::cout << "[MPI ADAPT] ranks=" << numRanks << ", energy=" << energy
              << std::endl;
    assert(std::fabs(energy - (-1.13)) < 1e-2 &&
           "MPI ADAPT energy does not match expected -1.13");
    assert(!ops.empty() && "Expected at least one operator to be selected");
    std::cout << "PASS" << std::endl;
  }

  cudaq::mpi::finalize();
  return 0;
}
